### DISCUSSION TITLE
Priority labels

### DISCUSSION THREAD

**Discussion by craigfowler** [elsa-workflows/elsa-core#702] [Design & tech]
Oops, had a bit of a conflict of priority labelling and I'll admit I was naughty to rename some. At the end of the day I don't really mind that much what priority labels we use.

## Do now, soon & later
Whenever I look at priority systems these days I always suggest the same three categories of priorities as a starting point:
1. Do now - jobs which interrupt us and must be done immediately, like a severe security vulnerability
2. Do soon - the next set of things on our roadmap
3. Do later - everything else which we might do, but we aren't ready to think about in detail yet

Actually, the reason for this dates back to old experience of mine. I worked with some product owners who would mark _almost everything_ in the project as high priority. I had to keep reminding them that _"If everything is high priority then really nothing is high priority"_.

I like the now/soon/later labels as a way of encouraging a different way of thinking about it. "Sure it has to be done, but _how soon_ does it have to be done? It's it the next thing we have to do or could we leave it 6 months?". If they mark everything as "do now" or even "do soon" then it becomes obvious that the team can't possibly do _all of that work_ in such a short time.

Of course, really they could be called 1, 2 & 3 or high, medium & low or anything else. Also, nobody on this project has exhibited that "planning anti-pattern" of pushing everything to high priority. It's just become habit for me but it's not necessarily needed here either.

So, whilst I altered the labels which had been created, I could just as happily change them back. Certainly high/medium/low priority are more recognizable than now/soon/later.

**Comment by jwillmer** [comment]
for me do-now sounds like it can only exist once and do-later I would never touch because "I do it later" 😁 

For prioritizing items within a sprint/backlog I find high and low prio important. 

maybe we could use some combination?:

- impact-severe (instead of do now)
- prio-high
- prio-low

since your label changes are related with the severity of a issue there mine are about a prio in the context of a collection of defined to-dos (milestone)

**Reply by jwillmer** [reply]
as an example:

a bug can have a high impact (for Elsa core 1) but a low priority since else 2 will address the issue

**Comment by craigfowler** [comment]
Although one of the examples I gave showed impact/severity, _I don't really mean these that way_. They 100% only relate to relative priority compared to the other issues on the tracker. I think actually that:

* Do soon means the exact same thing as priority: high
* Do later means the exact same thing as priority: low
* Do now is a new one which could mean priority: immediate

In fact, we could swap names back to these and it wouldn't make any difference to how I used them. High priority is still "things on the roadmap", low priority is still "things which aren't on the roadmap yet, but which we might want to put on the roadmap some day" and immediate priority is still the things which interrupt the roadmap and have to be done straight away.

Another example of a "do now" (or "do immediately", I just used "now" because it's a shorter word) issue is a blocker which stops is from doing things we want. Say we identify a refactor that we have to do before we can work on a group of 3 issues that we want to deliver in the current milestone.

Also, it's part of the plan that "do later" is ignored while you're doing normal work. Issues marked "do later" are the things in the backlog which aren't ready to work on yet. 
Thee only time you look at "do later" issues are when you're running out of issues in your milestones/roadmap and you're looking for ideas to bring forward into a new milestone. Once they are brought into a milestone, they are promoted to "do soon". As time moves on,  "later" becomes "soon", just like things which were lower priority a few months ago are higher priority now because "the things that were high priority a few months ago" are all completed now.

By getting used to ignoring the "do later" issues, it keeps keep us focused on the issues we want to deliver for the milestone. In fact I'd typically filter them out of search results so they don't fill up space in our heads when really we're not ready to be thinking about them.

Looking at your example of a high impact bug in Elsa 1.x. Actually by the time it's gone through triage I only see two likely outcomes from that.

* The issue is marked as "do now", if - during triage - we decide that we are 100% going to do an immediate hotfix release of Elsa 1.x
* The issue (for Elsa 1.x) is closed as a "wontfix"

I suppose _we could mark this hypothetical issue as "do later"_ if we can't decide whether we want to fix it in Elsa 1.x or not. One of the goals of triage though is to make that decision about priority. With a high impact bug, if you're going to leave it for a long time then probably that means you're never going to fix it.

The only other outcome I could imagine it's that a new/related issue is created as "do soon", which is to verify that the Elsa 1.x big doesn't exist in Elsa 2.

**Reply by jwillmer** [reply]
O.K. I can work with both label definitions but I think with respect to the audience we should keep the industry standards. Because it's not that customers will work on Elsa. It's developers that are used to the label names. What do you think?

**Reply by craigfowler** [reply]
Yeah, sure.  I agree with that.  As I mentioned, I realise it's a bias for me because I was burned a long time ago by a customer who needed some extra help to understand how to prioritise.

Elsa isn't that customer though, so when I think about it there's no good reason to use those unusual names.
I'll change the names back when I take my lunch & also amend the stale config.

**Reply by craigfowler** [reply]
Done, PR in for the stale config.
This evening what I will try to do is take a look at the issues marked stale and check that none of them needed any triage.
Mainly, that's checking that none of them were actually enhancement/bug issues (which never got given a label).

**Reply by jwillmer** [reply]
move your task to tomorrow since the bot waits 24h before he starts the first run ;)

**Reply by craigfowler** [reply]
Huh, I've already seen a number of notifications of stale bot already flashing a bundle of issues today. It's definitely run at least once already.

But tbh, I wasn't just going to search for stale labels, I was going to search for label-less issues.
